treeherder/perfalert/perfalert/__init__.py

```diff
@@ -55,104 +72,107 @@ def calc_t(w1, w2, weight_fn=None):
-        # Value of this point
-        self.value = value
 
-        # Which test run was this
-        self.testrun_id = testrun_id
-        # What revision this data is for
-        self.revision_id = revision_id
+        # What revision this data is for (usually, but not guaranteed
+        # to be increasing with push_timestamp)
+        self.push_id = push_id
+
+        # data values associated with this revision
+        self.values = copy.copy(values)
 
         # t-test score
         self.t = 0
+
         # Whether a perf regression is found
-        self.state = state
+        self.state = 'good'
 
     def __cmp__(self, o):
-        # only compare value to make sorting deterministic
-        # in cases where we have multiple datapoints with
-        # the same push timestamp
-        return cmp(
-            (self.push_timestamp, self.testrun_id, self.value),
-            (o.push_timestamp, o.testrun_id, o.value),
-        )
+        return self.push_timestamp < o.push_timestamp
 
     def __repr__(self):
-        return "<%s: %s, %s, %.3f, %.3f, %s>" % (self.push_timestamp,
-                                                 self.revision_id,
-                                                 self.testrun_id, self.value,
-                                                 self.t, self.state)
+        values_str = '[ %s ]' % ', '.join(['%.3f' % value for value in
+                                          self.values])
+        return "<%s: %s, %s, %.3f, %s>" % (self.push_timestamp,
+                                           self.push_id, values_str,
+                                           self.t, self.state)
 
 
 def detect_changes(data, min_back_window=12, max_back_window=24,
                    fore_window=12, t_threshold=7):
     # Use T-Tests
     # Analyze test data using T-Tests, comparing data[i-j:i] to data[i:i+k]
     data = sorted(data)
-    (j, k) = (min_back_window, fore_window)
-    good_data = []
 
-    num_points = len(data) - k + 1
     last_seen_regression = 0
-    for i in range(num_points):
+    for i in range(1, len(data)):
         di = data[i]
-        # if we haven't seen something that looks like a change in a
-        # while, incorporate some extra historical data into our t-test
-        # calculation
-        if last_seen_regression > min_back_window:
-            additional_back_window = min(last_seen_regression,
-                                         max_back_window - min_back_window)
-        else:
-            additional_back_window = 0
-        jw = [d.value for d in good_data[-1 * (j + additional_back_window):]]
-        kw = [d.value for d in data[i:i+k]]
 
-        # Reverse the backward data so that the current point is at the
-        # start of the window.
-        jw.reverse()
+        # keep on getting previous data until we've either got at least 12
+        # data points *or* we've hit the maximum back window
+        jw = []
+        di.amount_prev_data = 0
+        prev_indice = (i - 1)
```

Comment:
No, just an arbitrary decision from me. Probably not necessary, I'll take out the parans.